posthog/temporal/batch_exports/temporary_file.py

```diff
@@ -0,0 +1,523 @@
+        """
+        self._file.seek(0)
+        self._file.truncate()
+
+        self.bytes_since_last_reset = 0
+        self.records_since_last_reset = 0
+
+
+FlushCallable = collections.abc.Callable[
+    [BatchExportTemporaryFile, int, int, dt.datetime, bool], collections.abc.Awaitable[None]
+]
+
+
+class UnsupportedFileFormatError(Exception):
+    """Raised when a writer for an unsupported file format is requested."""
+
+    def __init__(self, file_format: str, destination: str):
+        super().__init__(f"{file_format} is not a supported format for {destination} batch exports.")
+
+
+class BatchExportWriter(abc.ABC):
+    """A temporary file writer to be used by batch export workflows.
+
+    Subclasses should define `_write_record_batch` with the particular intricacies
+    of the format they are writing as.
+
+    Actual writing calls are passed to the underlying `batch_export_file`.
+
+    Attributes:
+        _batch_export_file: The temporary file we are writing to.
+        max_bytes: Flush the temporary file with the provided `flush_callable`
+            upon reaching or surpassing this threshold. Keep in mind we write on a RecordBatch
+            per RecordBatch basis, which means the threshold will be surpassed by at most the
+            size of a RecordBatch before a flush occurs.
+        flush_callable: A callback to flush the temporary file when `max_bytes` is reached.
+            The temporary file will be reset after calling `flush_callable`.
+        file_kwargs: Optional keyword arguments passed when initializing `_batch_export_file`.
+        last_inserted_at: Latest `_inserted_at` written. This attribute leaks some implementation
+            details, as we are assuming assume `_inserted_at` is present, as it's added to all
+            batch export queries.
+        records_total: The total number of records (not RecordBatches!) written.
+        records_since_last_flush: The number of records written since last flush.
+        bytes_total: The total number of bytes written.
+        bytes_since_last_flush: The number of bytes written since last flush.
+    """
+
+    def __init__(
+        self,
+        flush_callable: FlushCallable,
+        max_bytes: int,
+        file_kwargs: collections.abc.Mapping[str, typing.Any] | None = None,
+    ):
+        self.flush_callable = flush_callable
+        self.max_bytes = max_bytes
+        self.file_kwargs: collections.abc.Mapping[str, typing.Any] = file_kwargs or {}
+
+        self._batch_export_file: BatchExportTemporaryFile | None = None
+        self.reset_writer_tracking()
+
+    def reset_writer_tracking(self):
+        """Reset this writer's tracking state."""
+        self.last_inserted_at: dt.datetime | None = None
+        self.records_total = 0
+        self.records_since_last_flush = 0
+        self.bytes_total = 0
+        self.bytes_since_last_flush = 0
+
+    @contextlib.asynccontextmanager
+    async def open_temporary_file(self):
```

Comment:
Hopefully this should make it clear when we are dealing with the writer and when the file: Inside this context, the file is opened and can be used. Once it's closed, file is no longer accessible, but the writer copies over tracking data for callers to inspect. File no longer being accessible is enforced by the last line clearing the reference we hold, and trying to access the file would raise a `ValueError` (there is a test case for this.